src/app/reporting/SynchronizedReportSchedulerImpl.cpp

```diff
@@ -180,25 +180,49 @@ CHIP_ERROR SynchronizedReportSchedulerImpl::CalculateNextReportTimeout(Timeout &
 /// the engine already verifies that read handlers are reportable before sending a report
 void SynchronizedReportSchedulerImpl::TimerFired()
 {
-    Timestamp now = mTimerDelegate->GetCurrentMonotonicTimestamp();
+    Timestamp now   = mTimerDelegate->GetCurrentMonotonicTimestamp();
+    bool firedEarly = true;
 
-    InteractionModelEngine::GetInstance()->GetReportingEngine().ScheduleRun();
-
-    mNodesPool.ForEachActiveObject([now](ReadHandlerNode * node) {
+    mNodesPool.ForEachActiveObject([now, &firedEarly](ReadHandlerNode * node) {
         if (node->GetMinTimestamp() <= now)
         {
             node->SetCanBeSynced(true);
         }
 
         if (node->IsReportableNow(now))
         {
+            // We set firedEarly false here because we assume we fired the timer early if no handler is reportable at the moment,
+            // which becomes false if we find a handler that is reportable
+            firedEarly = false;
             node->SetEngineRunScheduled(true);
             ChipLogProgress(DataManagement, "Handler: %p with min: 0x" ChipLogFormatX64 " and max: 0x" ChipLogFormatX64 "", (node),
                             ChipLogValueX64(node->GetMinTimestamp().count()), ChipLogValueX64(node->GetMaxTimestamp().count()));
         }
 
         return Loop::Continue;
     });
+
+    // If there are no handlers registers, no need to schedule the next report
+    if (mNodesPool.Allocated() && firedEarly)
+    {
+        ReadHandlerNode * firstNode = nullptr;
+
+        // If we fired the timer early, we need to restart the timer
+        mNodesPool.ForEachActiveObject([&firstNode](ReadHandlerNode * node) {
+            firstNode = node;
+            return Loop::Break;
+        });
+
+        VerifyOrReturn(firstNode); // Verify we have a valid node
+
+        Timeout timeout = Milliseconds32(0);
+        ReturnOnFailure(CalculateNextReportTimeout(timeout, firstNode, now));
```

Comment:
Could you add a comment of why the first node in the list is the correct/sufficient node to base time calculations on?

Reading the code without having too much background, I am wondering why calculateNextReportTimeout is based on the first node instead of being some "min" of all nodes.


Looking through the code, it seems the node is only used for ` VerifyOrReturnError(nullptr != FindReadHandlerNode(aNode->GetReadHandler()), CHIP_ERROR_INVALID_ARGUMENT)` ... which I do not understand. Why do we check for a read handler if aNode is not used at all after that?

Reply:
Looking at this: could you add comments specifically for what arguments in `ScheduleReport` actually do and mean? at least usage of `node` is unclear ot me

Reply:
The reason is that CalculateNextReportTimeout will iterate through all active nodes so it doesn't matter which one is passed, as long as it is a node for an existing readHandler.

The reason we do this check is that if for a reason the readHandler of a node just got freed and we start looping through all nodes, at some point IsReadHandlerReportable(node->GetReadHandler()) will be called, which yields:
IsReadHandlerReportable(null), null->ShouldStartReporting() -> crash. 

Now that being said, it is probably better to do null checks in all ReadHandler dependant functions of the Scheduler so I will do that instead.

Reply:
Updated, no more using the first node nor checking an unused variable.